readme.rst
Botan with PQ Schemes
=====================

This repository provides Botan integrated with some post-quantum schemes from the NIST PQC
competition. The implementations of post-quantum signature and public-key encryption schemes are
provided by `liboqs <https://github.com/open-quantum-safe/liboqs>`_

The modification were authored by Sebastian Ramacher and Simon Gärtner from AIT Austrian Institute
of Technology and are licensed under the same license as botan.

This work received funding from the Austrian Research Promotion Agency (FFG) under grant agreement
number FO999886370 (QKD4GOV).

The original description follows below.


Botan: Crypto and TLS for Modern C++
========================================

Botan (Japanese for peony flower) is a C++ cryptography library released under the
permissive `Simplified BSD <https://botan.randombit.net/license.txt>`_ license.

Botan's goal is to be the best option for cryptography in C++ by offering the
tools necessary to implement a range of practical systems, such as TLS protocol,
X.509 certificates, modern AEAD ciphers, PKCS#11 and TPM hardware support,
password hashing, and post quantum crypto schemes. A Python binding is included,
and several other `language bindings
<https://github.com/randombit/botan/wiki/Language-Bindings>`_ are available.
It is used in many `open source and commercial products <https://github.com/randombit/botan/wiki/Users>`_.
The library is accompanied by a featureful
`command line interface <https://botan.randombit.net/handbook/cli.html>`_.

See the `documentation <https://botan.randombit.net/handbook>`_ for more
information about included features.

Development is coordinated on `GitHub <https://github.com/randombit/botan>`_
and contributions are welcome. If you need help, please open an issue on
`GitHub <https://github.com/randombit/botan/issues>`_ or email the
`botan-devel mailing list <https://lists.randombit.net/mailman/listinfo/botan-devel/>`_.
New releases are announced on the `botan-announce mailing list
<https://lists.randombit.net/mailman/listinfo/botan-announce/>`_.
If you think you have found a security issue, see the `security page
<https://botan.randombit.net/security.html>`_ for contact information.

The latest release is
`2.19.2 <https://botan.randombit.net/releases/Botan-2.19.2.tar.xz>`_
`(sig) <https://botan.randombit.net/releases/Botan-2.19.2.tar.xz.asc>`_,
released on 2022-06-03.
All releases are signed with a `PGP key <https://botan.randombit.net/pgpkey.txt>`_.
See the `release notes <https://botan.randombit.net/news.html>`_ for
what is new. Botan is also available through most
`distributions <https://github.com/randombit/botan/wiki/Distros>`_
such as Fedora, Debian, Arch and Homebrew.

.. image:: https://api.travis-ci.com/randombit/botan.svg?branch=master
    :target: https://travis-ci.com/github/randombit/botan
    :alt: Travis CI status

.. image:: https://ci.appveyor.com/api/projects/status/n9f94dljd03j2lce/branch/master?svg=true
    :target: https://ci.appveyor.com/project/randombit/botan/branch/master
    :alt: AppVeyor CI status

.. image:: https://codecov.io/github/randombit/botan/coverage.svg?branch=master
    :target: https://codecov.io/github/randombit/botan
    :alt: Code coverage report

.. image:: https://img.shields.io/lgtm/alerts/g/randombit/botan.svg
    :target: https://lgtm.com/projects/g/randombit/botan/alerts/
    :alt: LGTM alerts

.. image:: https://oss-fuzz-build-logs.storage.googleapis.com/badges/botan.svg
    :target: https://oss-fuzz.com/coverage-report/job/libfuzzer_asan_botan/latest
    :alt: OSS-Fuzz status

.. image:: https://scan.coverity.com/projects/624/badge.svg
    :target: https://scan.coverity.com/projects/624
    :alt: Coverity results

.. image:: https://repology.org/badge/tiny-repos/botan.svg
    :target: https://repology.org/project/botan/versions
    :alt: Packaging status

.. image:: https://bestpractices.coreinfrastructure.org/projects/531/badge
    :target: https://bestpractices.coreinfrastructure.org/projects/531
    :alt: CII Best Practices statement

Find Enclosed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Transport Layer Security (TLS) Protocol
----------------------------------------

* TLS v1.2 and DTLS v1.2
* Supported extensions include session tickets, SNI, ALPN, OCSP stapling,
  encrypt-then-mac CBC, and extended master secret.
* Supports authentication using certificates or preshared keys (PSK)
* Supports record encryption with ChaCha20Poly1305, AES/OCB, AES/GCM, AES/CCM,
  Camellia/GCM as well as legacy CBC ciphersuites.
* Key exchange using CECPQ1, ECDH, FFDHE, or RSA

Public Key Infrastructure
----------------------------------------

* X.509v3 certificates and CRL creation and handling
* PKIX certificate path validation, including name constraints.
* OCSP request creation and response handling
* PKCS #10 certificate request generation and processing
* Access to Windows, macOS and Unix system certificate stores
* SQL database backed certificate store

Public Key Cryptography
----------------------------------------

* RSA signatures and encryption
* DH and ECDH key agreement
* Signature schemes ECDSA, DSA, Ed25519, ECGDSA, ECKCDSA, SM2, GOST 34.10
* Post-quantum signature scheme XMSS
* Post-quantum key agreement schemes McEliece and NewHope
* ElGamal encryption
* Padding schemes OAEP, PSS, PKCS #1 v1.5, X9.31

Ciphers, hashes, MACs, and checksums
----------------------------------------

* Authenticated cipher modes EAX, OCB, GCM, SIV, CCM, (X)ChaCha20Poly1305
* Cipher modes CTR, CBC, XTS, CFB, OFB
* Block ciphers AES, ARIA, Blowfish, Camellia, CAST-128, DES/3DES, IDEA,
  Lion, SEED, Serpent, SHACAL2, SM4, Threefish-512, Twofish
* Stream ciphers (X)ChaCha20, (X)Salsa20, SHAKE-128, RC4
* Hash functions SHA-1, SHA-2, SHA-3, MD5, RIPEMD-160, BLAKE2b,
  Skein-512, SM3, Streebog, Whirlpool
* Authentication codes HMAC, CMAC, Poly1305, SipHash, GMAC, X9.19 DES-MAC
* Non-cryptographic checksums Adler32, CRC24, CRC32

Other Useful Things
----------------------------------------

* Full C++ PKCS #11 API wrapper
* Interfaces for TPM v1.2 device access
* Simple compression API wrapping zlib, bzip2, and lzma libraries
* RNG wrappers for system RNG and hardware RNGs
* HMAC_DRBG and entropy collection system for userspace RNGs
* Password hashing schemes PBKDF2, Argon2, Scrypt, bcrypt
* SRP-6a password authenticated key exchange
* Key derivation functions including HKDF, KDF2, SP 800-108, SP 800-56A, SP 800-56C
* HOTP and TOTP algorithms
* Format preserving encryption scheme FE1
* Threshold secret sharing
* NIST key wrapping
* Boost.Asio compatible TLS client stream


news.rst
Release Notes
========================================

Version 3.0.0, Not Yet Released
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* Botan is now a C++20 codebase; compiler requirements have been
  increased to GCC 11, Clang 14, or MSVC 2022. (GH #2455 #3086)

Breaking Changes
----------------------------------------

* Remove many deprecated headers. In particular all algorithm specific
  headers (such as ``aes.h``) are no longer available; instead objects
  must be created through the base class ``create`` functions. (GH #2456)

* Removed most functions previously marked as deprecated.

* Remove several deprecated algorithms including CAST-256, MISTY1, Kasumi,
  DESX, XTEA, PBKDF1, MCEIES, CBC-MAC, Tiger, NEWHOPE, and CECPQ1 (GH #2434 #3094)

* Remove the entropy source which walked ``/proc`` as it is no longer
  required on modern systems. (GH #2692)

* Remove the entropy source which reads from ``/dev/random`` as it is
  supplanted by the extant source one which reads from the system RNG.
  (GH #2636)

* Remove use of ``shared_ptr`` from certificate store API, as since
  2.4.0 ``X509_Certificate`` is internally a ``shared_ptr``. (GH #2484)

* Several enums including ``DL_Group::Format``, ``EC_Group_Formatting``,
  ``CRL_Code``, ``ASN1_Tag``, ``Key_Constraints`` and ``Signature_Format`` are
  now ``enum class``.  The ``ASN1_Tag`` enum has been split into ``ASN1_Type``
  and ``ASN1_Class``.  (GH #2551 #2552 #3084 #2584 #3225)

* Avoid using or returning raw pointers whenever possible. (GH #2683 #2684
  #2685 #2687 #2688 #2690 #2691 #2693 #2694 #2695 #2696 #2697 #2700 #2703 #2708
  #3220)

* Remove support for HP and Pathscale compilers, Google NaCL (GH #2455),
  and IncludeOS (GH #3406)

* Remove deprecated ``Data_Store`` class (GH #2461)

* Remove deprecated public member variables of ``OID``, ``Attribute``,
  ``BER_Object``, and ``AlgorithmIdentifier``. (GH #2462)

* "SHA-160" and "SHA1" are no longer recognized as names for "SHA-1"
  (GH #3186)

TLS Changes
----------------------------------------

* Added support for TLS v1.3

* Support for TLS 1.0, TLS 1.1, and DTLS 1.0 have been removed (GH #2631)

* Remove several deprecated features in TLS including DSA ciphersuites (GH #2505),
  anonymous ciphersuites (GH #2497), SHA-1 signatures in TLS 1.2 (GH #2537),
  SRP ciphersuites (GH #2506), SEED ciphersuites (GH #2509),
  Camellia CBC ciphersuites (GH #2509), AES-128 OCB ciphersuites (GH #2511),
  DHE_PSK suites (GH #2512), CECPQ1 ciphersuites (GH #3094)

New Cryptographic Algorithms
----------------------------------------

* Add support for Kyber post-quantum KEM (GH #2872 #2500)

* Add support for Dilithium lattice based signatures (GH #2973 #3212)

* Add support for hashing onto an elliptic curve using the SSWU
  technique of draft-irtf-cfrg-hash-to-curve (GH #2726)

* Add support for keyed BLAKE2b (GH #2524)

New APIs
----------------------------------------

* Add new interface ``T::new_object`` which supplants ``T::clone``. The
  difference is that ``new_object`` returns a ``unique_ptr<T>`` instead of a raw
  pointer ``T*``. ``T::clone`` is retained but simply releases the result of
  ``new_object``. (GH #2689 #2704)

* Add an API to ``PasswordHash`` accepting an AD and/or secret key, allowing
  those facilities to be used without using an algorithm specific API (GH #2707)

* Add new ``X509_DN::DER_encode`` function. (GH #2472)

* New API ``Public_Key::get_int_field`` for getting the integer fields of a public
  (or private) key by name (GH #3200)

* New ``Cipher_Mode`` APIs ``ideal_granularity`` and ``requires_entire_message``
  (GH #3172 #3168)


src/lib/pubkey/pk_algs.cpp
/*
* PK Key
* (C) 1999-2010,2016 Jack Lloyd
*
* Botan is released under the Simplified BSD License (see license.txt)
*/

#include <botan/pk_algs.h>
#include <botan/internal/parsing.h>

#if defined(BOTAN_HAS_RSA)
  #include <botan/rsa.h>
#endif

#if defined(BOTAN_HAS_DSA)
  #include <botan/dsa.h>
#endif

#if defined(BOTAN_HAS_DL_GROUP)
  #include <botan/dl_group.h>
#endif

#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
  #include <botan/dh.h>
#endif

#if defined(BOTAN_HAS_ECC_PUBLIC_KEY_CRYPTO)
  #include <botan/ecc_key.h>
#endif

#if defined(BOTAN_HAS_ECDSA)
  #include <botan/ecdsa.h>
#endif

#if defined(BOTAN_HAS_ECGDSA)
  #include <botan/ecgdsa.h>
#endif

#if defined(BOTAN_HAS_ECKCDSA)
  #include <botan/eckcdsa.h>
#endif

#if defined(BOTAN_HAS_ED25519)
  #include <botan/ed25519.h>
#endif

#if defined(BOTAN_HAS_GOST_34_10_2001)
  #include <botan/gost_3410.h>
#endif

#if defined(BOTAN_HAS_ELGAMAL)
  #include <botan/elgamal.h>
#endif

#if defined(BOTAN_HAS_ECDH)
  #include <botan/ecdh.h>
#endif

#if defined(BOTAN_HAS_CURVE_25519)
  #include <botan/curve25519.h>
#endif

#if defined(BOTAN_HAS_MCELIECE)
  #include <botan/mceliece.h>
#endif

#if defined(BOTAN_HAS_XMSS_RFC8391)
  #include <botan/xmss.h>
#endif

#if defined(BOTAN_HAS_SM2)
  #include <botan/sm2.h>
#endif

#if defined(BOTAN_HAS_OQS)
#include <botan/oqs-integration.h>
#endif

namespace Botan {

std::unique_ptr<Public_Key>
load_public_key(const AlgorithmIdentifier& alg_id,
                const std::vector<uint8_t>& key_bits)
   {
   const std::string oid_str = alg_id.get_oid().to_formatted_string();
   const std::vector<std::string> alg_info = split_on(oid_str, '/');
   const std::string alg_name = alg_info[0];

#if defined(BOTAN_HAS_RSA)
   if(alg_name == "RSA")
      return std::make_unique<RSA_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_CURVE_25519)
   if(alg_name == "Curve25519")
      return std::make_unique<Curve25519_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_MCELIECE)
   if(alg_name == "McEliece")
      return std::make_unique<McEliece_PublicKey>(key_bits);
#endif

#if defined(BOTAN_HAS_ECDSA)
   if(alg_name == "ECDSA")
      return std::make_unique<ECDSA_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ECDH)
   if(alg_name == "ECDH")
      return std::make_unique<ECDH_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
   if(alg_name == "DH")
      return std::make_unique<DH_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_DSA)
   if(alg_name == "DSA")
      return std::make_unique<DSA_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ELGAMAL)
   if(alg_name == "ElGamal")
      return std::make_unique<ElGamal_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ECGDSA)
   if(alg_name == "ECGDSA")
      return std::make_unique<ECGDSA_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ECKCDSA)
   if(alg_name == "ECKCDSA")
      return std::make_unique<ECKCDSA_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ED25519)
   if(alg_name == "Ed25519")
      return std::make_unique<Ed25519_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_GOST_34_10_2001)
   if(alg_name == "GOST-34.10" || alg_name == "GOST-34.10-2012-256" || alg_name == "GOST-34.10-2012-512")
      return std::make_unique<GOST_3410_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_SM2)
   if(alg_name == "SM2" || alg_name == "SM2_Sig" || alg_name == "SM2_Enc")
      return std::make_unique<SM2_PublicKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_XMSS_RFC8391)
   if(alg_name == "XMSS")
      return std::make_unique<XMSS_PublicKey>(key_bits);
#endif

#if defined(BOTAN_HAS_OQS)
    if (alg_name == "Dilithium 2" || alg_name == "Dilithium 3" || alg_name == "Dilithium 5" ||
        alg_name == "Dilithium 2 AES" || alg_name == "Dilithium 3 AES" ||
        alg_name == "Dilithium 5 AES" || alg_name == "Picnic L1 FS" ||
        alg_name == "Picnic L1 Full" || alg_name == "Picnic L3 FS" ||
        alg_name == "Picnic L3 Full" || alg_name == "Picnic L5 FS" ||
        alg_name == "Picnic L5 Full" || alg_name == "Picnic3 L1" || alg_name == "Picnic3 L3" ||
        alg_name == "Picnic3 L5" || alg_name == "SPHINCS+-Haraka-128f-robust" ||
        alg_name == "SPHINCS+-Haraka-128f-simple" || alg_name == "SPHINCS+-Haraka-128s-robust" ||
        alg_name == "SPHINCS+-Haraka-128s-simple" || alg_name == "SPHINCS+-Haraka-192f-robust" ||
        alg_name == "SPHINCS+-Haraka-192f-simple" || alg_name == "SPHINCS+-Haraka-192s-robust" ||
        alg_name == "SPHINCS+-Haraka-192s-simple" || alg_name == "SPHINCS+-Haraka-256f-robust" ||
        alg_name == "SPHINCS+-Haraka-256f-simple" || alg_name == "SPHINCS+-Haraka-256s-robust" ||
        alg_name == "SPHINCS+-Haraka-256s-simple" || alg_name == "SPHINCS+-SHA256-128f-robust" ||
        alg_name == "SPHINCS+-SHA256-128f-simple" || alg_name == "SPHINCS+-SHA256-128s-robust" ||
        alg_name == "SPHINCS+-SHA256-128s-simple" || alg_name == "SPHINCS+-SHA256-192f-robust" ||
        alg_name == "SPHINCS+-SHA256-192f-simple" || alg_name == "SPHINCS+-SHA256-192s-robust" ||
        alg_name == "SPHINCS+-SHA256-192s-simple" || alg_name == "SPHINCS+-SHA256-256f-robust" ||
        alg_name == "SPHINCS+-SHA256-256f-simple" || alg_name == "SPHINCS+-SHA256-256s-robust" ||
        alg_name == "SPHINCS+-SHA256-256s-simple" || alg_name == "SPHINCS+-SHAKE256-128f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-128f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-128s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-128s-simple" ||
        alg_name == "SPHINCS+-SHAKE256-192f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-192f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-192s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-192s-simple" ||
        alg_name == "SPHINCS+-SHAKE256-256f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-256f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-256s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-256s-simple") {
      return std::make_unique<PQ_PublicKey>(alg_id, key_bits);
    }
#endif

    throw Decoding_Error("Unknown or unavailable public key algorithm " + alg_name);
  }

  std::unique_ptr<Private_Key> load_private_key(const AlgorithmIdentifier& alg_id,
                                                const secure_vector<uint8_t>& key_bits) {
    const std::string alg_name = alg_id.get_oid().to_formatted_string();

#if defined(BOTAN_HAS_RSA)
   if(alg_name == "RSA")
      return std::make_unique<RSA_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_CURVE_25519)
   if(alg_name == "Curve25519")
      return std::make_unique<Curve25519_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ECDSA)
   if(alg_name == "ECDSA")
      return std::make_unique<ECDSA_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ECDH)
   if(alg_name == "ECDH")
      return std::make_unique<ECDH_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
   if(alg_name == "DH")
      return std::make_unique<DH_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_DSA)
   if(alg_name == "DSA")
      return std::make_unique<DSA_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_MCELIECE)
   if(alg_name == "McEliece")
      return std::make_unique<McEliece_PrivateKey>(key_bits);
#endif

#if defined(BOTAN_HAS_ECGDSA)
   if(alg_name == "ECGDSA")
      return std::make_unique<ECGDSA_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ECKCDSA)
   if(alg_name == "ECKCDSA")
      return std::make_unique<ECKCDSA_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ED25519)
   if(alg_name == "Ed25519")
      return std::make_unique<Ed25519_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_GOST_34_10_2001)
   if(alg_name == "GOST-34.10" || alg_name == "GOST-34.10-2012-256" || alg_name == "GOST-34.10-2012-512")
      return std::make_unique<GOST_3410_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_SM2)
   if(alg_name == "SM2" || alg_name == "SM2_Sig" || alg_name == "SM2_Enc")
      return std::make_unique<SM2_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_ELGAMAL)
   if(alg_name == "ElGamal")
      return std::make_unique<ElGamal_PrivateKey>(alg_id, key_bits);
#endif

#if defined(BOTAN_HAS_XMSS_RFC8391)
   if(alg_name == "XMSS")
      return std::make_unique<XMSS_PrivateKey>(key_bits);
#endif

#if defined(BOTAN_HAS_OQS)
    if (alg_name == "Dilithium 2" || alg_name == "Dilithium 3" || alg_name == "Dilithium 5" ||
        alg_name == "Dilithium 2 AES" || alg_name == "Dilithium 3 AES" ||
        alg_name == "Dilithium 5 AES" || alg_name == "Picnic L1 FS" ||
        alg_name == "Picnic L1 Full" || alg_name == "Picnic L3 FS" ||
        alg_name == "Picnic L3 Full" || alg_name == "Picnic L5 FS" ||
        alg_name == "Picnic L5 Full" || alg_name == "Picnic3 L1" || alg_name == "Picnic3 L3" ||
        alg_name == "Picnic3 L5" || alg_name == "SPHINCS+-Haraka-128f-robust" ||
        alg_name == "SPHINCS+-Haraka-128f-simple" || alg_name == "SPHINCS+-Haraka-128s-robust" ||
        alg_name == "SPHINCS+-Haraka-128s-simple" || alg_name == "SPHINCS+-Haraka-192f-robust" ||
        alg_name == "SPHINCS+-Haraka-192f-simple" || alg_name == "SPHINCS+-Haraka-192s-robust" ||
        alg_name == "SPHINCS+-Haraka-192s-simple" || alg_name == "SPHINCS+-Haraka-256f-robust" ||
        alg_name == "SPHINCS+-Haraka-256f-simple" || alg_name == "SPHINCS+-Haraka-256s-robust" ||
        alg_name == "SPHINCS+-Haraka-256s-simple" || alg_name == "SPHINCS+-SHA256-128f-robust" ||
        alg_name == "SPHINCS+-SHA256-128f-simple" || alg_name == "SPHINCS+-SHA256-128s-robust" ||
        alg_name == "SPHINCS+-SHA256-128s-simple" || alg_name == "SPHINCS+-SHA256-192f-robust" ||
        alg_name == "SPHINCS+-SHA256-192f-simple" || alg_name == "SPHINCS+-SHA256-192s-robust" ||
        alg_name == "SPHINCS+-SHA256-192s-simple" || alg_name == "SPHINCS+-SHA256-256f-robust" ||
        alg_name == "SPHINCS+-SHA256-256f-simple" || alg_name == "SPHINCS+-SHA256-256s-robust" ||
        alg_name == "SPHINCS+-SHA256-256s-simple" || alg_name == "SPHINCS+-SHAKE256-128f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-128f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-128s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-128s-simple" ||
        alg_name == "SPHINCS+-SHAKE256-192f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-192f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-192s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-192s-simple" ||
        alg_name == "SPHINCS+-SHAKE256-256f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-256f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-256s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-256s-simple") {
      return std::make_unique<PQ_PrivateKey>(alg_id, key_bits);
    }
#endif

    throw Decoding_Error("Unknown or unavailable public key algorithm " + alg_name);
  }

#if defined(BOTAN_HAS_ECC_GROUP)

  namespace {

    std::string default_ec_group_for(const std::string& alg_name) {
      if (alg_name == "SM2" || alg_name == "SM2_Enc" || alg_name == "SM2_Sig")
        return "sm2p256v1";
      if (alg_name == "GOST-34.10" || alg_name == "GOST-34.10-2012-256")
        return "gost_256A";
      if (alg_name == "GOST-34.10-2012-512")
        return "gost_512A";
      if (alg_name == "ECGDSA")
        return "brainpool256r1";
      return "secp256r1";
    }

  } // namespace

#endif

BOTAN_PUBLIC_API(3,0) std::unique_ptr<Private_Key>
create_ec_private_key(const std::string& alg_name,
                      const EC_Group& ec_group,
                      RandomNumberGenerator& rng)
   {
#if defined(BOTAN_HAS_ECDSA)
   if(alg_name == "ECDSA")
      return std::make_unique<ECDSA_PrivateKey>(rng, ec_group);
#endif

#if defined(BOTAN_HAS_ECDH)
   if(alg_name == "ECDH")
      return std::make_unique<ECDH_PrivateKey>(rng, ec_group);
#endif

#if defined(BOTAN_HAS_ECKCDSA)
   if(alg_name == "ECKCDSA")
      return std::make_unique<ECKCDSA_PrivateKey>(rng, ec_group);
#endif

#if defined(BOTAN_HAS_GOST_34_10_2001)
   if(alg_name == "GOST-34.10" || alg_name == "GOST-34.10-2012-256" || alg_name == "GOST-34.10-2012-512")
      return std::make_unique<GOST_3410_PrivateKey>(rng, ec_group);
#endif

#if defined(BOTAN_HAS_SM2)
   if(alg_name == "SM2" || alg_name == "SM2_Sig" || alg_name == "SM2_Enc")
      return std::make_unique<SM2_PrivateKey>(rng, ec_group);
#endif

#if defined(BOTAN_HAS_ECGDSA)
   if(alg_name == "ECGDSA")
      return std::make_unique<ECGDSA_PrivateKey>(rng, ec_group);
#endif

   return nullptr;
   }


std::unique_ptr<Private_Key>
create_private_key(const std::string& alg_name,
                   RandomNumberGenerator& rng,
                   const std::string& params,
                   const std::string& provider)
   {
   /*
   * Default paramaters are chosen for work factor > 2**128 where possible
   */

#if defined(BOTAN_HAS_CURVE_25519)
   if(alg_name == "Curve25519")
      return std::make_unique<Curve25519_PrivateKey>(rng);
#endif

#if defined(BOTAN_HAS_RSA)
   if(alg_name == "RSA")
      {
      const size_t rsa_bits = (params.empty() ? 3072 : to_u32bit(params));
      return std::make_unique<RSA_PrivateKey>(rng, rsa_bits);
      }
#endif

#if defined(BOTAN_HAS_MCELIECE)
   if(alg_name == "McEliece")
      {
      std::vector<std::string> mce_param =
         Botan::split_on(params.empty() ? "2960,57" : params, ',');

      if(mce_param.size() != 2)
         throw Invalid_Argument("create_private_key bad McEliece parameters " + params);

      size_t mce_n = Botan::to_u32bit(mce_param[0]);
      size_t mce_t = Botan::to_u32bit(mce_param[1]);

      return std::make_unique<Botan::McEliece_PrivateKey>(rng, mce_n, mce_t);
      }
#endif

#if defined(BOTAN_HAS_XMSS_RFC8391)
   if(alg_name == "XMSS")
      {
      return std::make_unique<XMSS_PrivateKey>(XMSS_Parameters(params.empty() ? "XMSS-SHA2_10_512" : params).oid(), rng);
      }
#endif

#if defined(BOTAN_HAS_ED25519)
   if(alg_name == "Ed25519")
      {
      return std::make_unique<Ed25519_PrivateKey>(rng);
      }
#endif

   // ECC crypto
#if defined(BOTAN_HAS_ECC_PUBLIC_KEY_CRYPTO)

   if(alg_name == "ECDSA" ||
      alg_name == "ECDH" ||
      alg_name == "ECKCDSA" ||
      alg_name == "ECGDSA" ||
      alg_name == "SM2" ||
      alg_name == "SM2_Sig" ||
      alg_name == "SM2_Enc" ||
      alg_name == "GOST-34.10" ||
      alg_name == "GOST-34.10-2012-256" ||
      alg_name == "GOST-34.10-2012-512")
      {
      const EC_Group ec_group(params.empty() ? default_ec_group_for(alg_name) : params);
      return create_ec_private_key(alg_name, ec_group, rng);
      }
#endif

   // DL crypto
#if defined(BOTAN_HAS_DL_GROUP)
   if(alg_name == "DH" || alg_name == "DSA" || alg_name == "ElGamal")
      {
      std::string default_group = (alg_name == "DSA") ? "dsa/botan/2048" : "modp/ietf/2048";
      DL_Group modp_group(params.empty() ? default_group : params);

#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
      if(alg_name == "DH")
         return std::make_unique<DH_PrivateKey>(rng, modp_group);
#endif

#if defined(BOTAN_HAS_DSA)
      if(alg_name == "DSA")
         return std::make_unique<DSA_PrivateKey>(rng, modp_group);
#endif

#if defined(BOTAN_HAS_ELGAMAL)
      if(alg_name == "ElGamal")
         return std::make_unique<ElGamal_PrivateKey>(rng, modp_group);
#endif
      }
#endif

#if defined(BOTAN_HAS_OQS)
    if (alg_name == "Dilithium 2" || alg_name == "Dilithium 3" || alg_name == "Dilithium 5" ||
        alg_name == "Dilithium 2 AES" || alg_name == "Dilithium 3 AES" ||
        alg_name == "Dilithium 5 AES" || alg_name == "Picnic L1 FS" ||
        alg_name == "Picnic L1 Full" || alg_name == "Picnic L3 FS" ||
        alg_name == "Picnic L3 Full" || alg_name == "Picnic L5 FS" ||
        alg_name == "Picnic L5 Full" || alg_name == "Picnic3 L1" || alg_name == "Picnic3 L3" ||
        alg_name == "Picnic3 L5" || alg_name == "SPHINCS+-Haraka-128f-robust" ||
        alg_name == "SPHINCS+-Haraka-128f-simple" || alg_name == "SPHINCS+-Haraka-128s-robust" ||
        alg_name == "SPHINCS+-Haraka-128s-simple" || alg_name == "SPHINCS+-Haraka-192f-robust" ||
        alg_name == "SPHINCS+-Haraka-192f-simple" || alg_name == "SPHINCS+-Haraka-192s-robust" ||
        alg_name == "SPHINCS+-Haraka-192s-simple" || alg_name == "SPHINCS+-Haraka-256f-robust" ||
        alg_name == "SPHINCS+-Haraka-256f-simple" || alg_name == "SPHINCS+-Haraka-256s-robust" ||
        alg_name == "SPHINCS+-Haraka-256s-simple" || alg_name == "SPHINCS+-SHA256-128f-robust" ||
        alg_name == "SPHINCS+-SHA256-128f-simple" || alg_name == "SPHINCS+-SHA256-128s-robust" ||
        alg_name == "SPHINCS+-SHA256-128s-simple" || alg_name == "SPHINCS+-SHA256-192f-robust" ||
        alg_name == "SPHINCS+-SHA256-192f-simple" || alg_name == "SPHINCS+-SHA256-192s-robust" ||
        alg_name == "SPHINCS+-SHA256-192s-simple" || alg_name == "SPHINCS+-SHA256-256f-robust" ||
        alg_name == "SPHINCS+-SHA256-256f-simple" || alg_name == "SPHINCS+-SHA256-256s-robust" ||
        alg_name == "SPHINCS+-SHA256-256s-simple" || alg_name == "SPHINCS+-SHAKE256-128f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-128f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-128s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-128s-simple" ||
        alg_name == "SPHINCS+-SHAKE256-192f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-192f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-192s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-192s-simple" ||
        alg_name == "SPHINCS+-SHAKE256-256f-robust" ||
        alg_name == "SPHINCS+-SHAKE256-256f-simple" ||
        alg_name == "SPHINCS+-SHAKE256-256s-robust" ||
        alg_name == "SPHINCS+-SHAKE256-256s-simple") {
      return std::make_unique<PQ_PrivateKey>(alg_name);
    }
#endif

    BOTAN_UNUSED(alg_name, rng, params, provider);

    return std::unique_ptr<Private_Key>();
  }

  std::vector<std::string> probe_provider_private_key(const std::string& alg_name,
                                                      const std::vector<std::string>& possible) {
    std::vector<std::string> providers;

    for (auto&& prov : possible) {
      if (prov == "base")
        providers.push_back(prov);
    }

    BOTAN_UNUSED(alg_name);

    return providers;
  }
} // namespace Botan
